File: app/src/test/java/com/test/xyz/demo/ui/repolist/mvp/RepoListPresenterTest.java
package com.test.xyz.demo.ui.repolist.mvp;

import com.test.xyz.demo.domain.interactor.MainInteractor;
import com.test.xyz.demo.ui.BasePresenterTest;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

public class RepoListPresenterTest extends BasePresenterTest {
    private static final String USER_NAME = "google";

    private RepoListPresenter repoListPresenter;

    @Mock
    private RepoListView repoListView;

    @Mock
    private MainInteractor mainInteractor;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        mockInteractor(mainInteractor);
        repoListPresenter = new RepoListPresenterImpl(repoListView, mainInteractor);
    }

    @Test
    public void requestRepoList_shouldReturnRepoList() {
        //GIVEN
        //NOTHING

        //WHEN
        repoListPresenter.requestRepoList(USER_NAME);

        //THEN
        verify(repoListView).showRepoList(any(List.class));
        verify(repoListView, never()).showError(any(String.class));
    }

    @Test
    public void requestRepoList_whenUserNameIsEmpty_shouldError() {
        //GIVEN
        //NOTHING

        //WHEN
        repoListPresenter.requestRepoList("");

        //THEN
        verify(repoListView, never()).showRepoList(any(List.class));
        verify(repoListView).showError(any(String.class));
    }
}

File: app/src/test/java/com/test/xyz/demo/ui/repolist/vp/RepoListPresenterTest.kt
@file:Suppress("IllegalIdentifier")
package com.test.xyz.demo.ui.repolist.vp

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.never
import com.nhaarman.mockito_kotlin.verify
import com.test.xyz.demo.domain.interactor.MainInteractor
import com.test.xyz.demo.ui.BasePresenterTest
import org.junit.Before
import org.junit.Test

class RepoListPresenterTest : BasePresenterTest() {

    lateinit var repoListPresenter: RepoListPresenter

    val repoListView: RepoListView = mock()
    val mainInteractor: MainInteractor = mock()

    @Before
    fun setup() {
        mockInteractor(mainInteractor)
        repoListPresenter = RepoListPresenterImpl(repoListView, mainInteractor)
    }

    @Test
    fun `requestRepoList should return RepoList`() {
        //GIVEN
        //NOTHING

        //WHEN
        repoListPresenter.requestRepoList(USER_NAME)

        //THEN
        verify(repoListView).showRepoList(any())
        verify(repoListView, never()).showError(any<String>())
    }

    @Test
    fun `requestRepoList when UserName is empty should return Error`() {
        //GIVEN
        //NOTHING

        //WHEN
        repoListPresenter.requestRepoList("")

        //THEN
        verify(repoListView, never()).showRepoList(any())
        verify(repoListView).showError(any<String>())
    }

    companion object {
        private val USER_NAME = "google"
    }
}
